grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandlerBuilder.java

```diff
@@ -0,0 +1,221 @@
+     * result in an {@link IllegalArgumentException}.
+     *
+     * @param messageTypes The message types to register
+     */
+    @SafeVarargs
+    public final UnframedGrpcErrorHandlerBuilder registerMarshalledMessages(
+            Class<? extends Message>... messageTypes) {
+        requireNonNull(messageTypes, "messageTypes");
+        if (jsonMarshaller != null) {
+            throw new IllegalArgumentException(
+                    "Cannot register custom messageTypes because a custom JSON marshaller has already been " +
+                    "set. Use the custom marshaller to register custom message types.");
+        }
+        this.messageTypes.addAll(ImmutableList.copyOf(messageTypes));
+        return this;
+    }
+
+    /**
+     * Registers custom messages or message types to be marshalled by the error handler.
+     *
+     * <p>This method allows registering either message instances or message types for
+     * custom error responses. Each element in {@code messagesOrMessageTypes} must either
+     * be an instance of the {@link Message} class　or a class that extends the {@link Message}
+     * class. If a custom JSON marshaller has already been set,
+     * calling this method will result in an {@link IllegalArgumentException}.
+     *
+     * @param messagesOrMessageTypes The messages or message types to register
+     */
+    public UnframedGrpcErrorHandlerBuilder registerMarshalledMessages(Iterable<?> messagesOrMessageTypes) {
+        requireNonNull(messagesOrMessageTypes, "messagesOrMessageTypes");
+        if (jsonMarshaller != null) {
+            throw new IllegalArgumentException(
+                    "Cannot register custom messagesOrMessageTypes because a custom JSON marshaller has " +
+                    "already been set. Use the custom marshaller to register custom messages or message types."
+            );
+        }
+
+        for (final Object messageOrMessageType : messagesOrMessageTypes) {
+            requireNonNull(messageOrMessageType, "messagesOrMessageTypes contains null.");
+            if (messageOrMessageType instanceof Message) {
+                messages.add((Message) messageOrMessageType);
+            } else if (messageOrMessageType instanceof Class &&
+                       Message.class.isAssignableFrom((Class<?>) messageOrMessageType)) {
+                messageTypes.add((Class<? extends Message>) messageOrMessageType);
+            } else {
+                final String className =
+                        messageOrMessageType instanceof Class ? ((Class<?>) messageOrMessageType).getName()
+                                                              : messageOrMessageType.getClass().getName();
+                throw new IllegalArgumentException(
+                        className + " is neither Message class nor Message subclass.");
+            }
+        }
+
+        return this;
+    }
+
+    /**
+     * Builds and returns an instance of {@link UnframedGrpcErrorHandler}.
```

Comment:
Could be simplified like this?

```suggestion
     * Returns a newly created {@link UnframedGrpcErrorHandler}.
```